subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java

```diff
@@ -494,8 +487,20 @@ private TriFunction<com.microsoft.z3.Expr, Model, List<Decl<?>>, Expr<?>> exprBi
         return (term, model, vars) -> {
             final com.microsoft.z3.Expr[] args = term.getArgs();
             checkArgument(args.length == 2, "Number of arguments must be two");
-            final Expr<?> op1 = transform(args[0], model, vars);
-            final Expr<?> op2 = transform(args[1], model, vars);
+            Expr<?> op1;
+            Expr<?> op2;
+            boolean isInverse = false;
+            try {
+                op1 = transform(args[0], model, vars);
+            } catch (Exception e) {
+                op1 = transform(args[1], model, vars);
+                isInverse = true;
+            }
+            try {
+                op2 = transform(args[isInverse ? 0 : 1], model, vars);
+            } catch (Exception e) {
+                op2 = transform(args[isInverse ? 0 : 1], model, vars, op1.getType());
+            }
```

Comment:
I strongly recommend against this pattern, as it completely shatters the control flow. Moreover, catching `Exception` should not happen in code, as it also catches all of the validations we do. Why is this needed exactly? I am sure we can come up with a better solution.

Reply:
The correct commit is 6fd805f